carma-messenger-core/cpp_message/src/SDSM_Message.cpp

```diff
@@ -1072,4 +1072,958 @@ namespace cpp_message
 
     }
 
+
+
+    // SDSM decoding
+    boost::optional<j3224_v2x_msgs::msg::SensorDataSharingMessage> SDSM_Message::decode_sdsm_message(std::vector<uint8_t>&binary_array){
+        
+        j3224_v2x_msgs::msg::SensorDataSharingMessage output;
+        // decode results - stored in binary_array
+        asn_dec_rval_t rval;
+        MessageFrame_t* message = nullptr;
+        
+        // copy from vector to array         
+        size_t len = binary_array.size();    
+        uint8_t buf[len];             
+        std::copy(binary_array.begin(),binary_array.end(),buf);
+        
+        // use asn1c lib to decode
+        rval = uper_decode(0, &asn_DEF_MessageFrame,(void **) &message, buf, len, 0, 0);
+
+        if(rval.code == RC_OK)
+        {
+            // Incoming SDSM in ASN.1 C-struct format
+            SensorDataSharingMessage_t sdsm_core = message->value.choice.SensorDataSharingMessage;
+            
+            // MessageCount
+            if(sdsm_core.msgCnt < j2735_v2x_msgs::msg::MsgCount::MSG_COUNT_MAX){
+                output.msg_cnt.msg_cnt = sdsm_core.msgCnt;
+                std::cout << sdsm_core.msgCnt << std::endl;
+            }
+            else{
+                RCLCPP_WARN_STREAM(node_logging_->get_logger(), "Decoded SDSM MsgCount above max");
+            }
+
+            // SourceID
+            if(sdsm_core.sourceID.size > 0){
+                auto id_len = sdsm_core.sourceID.size;
+                for(auto i = 0; i < id_len; i++){
+                    output.source_id.id.push_back(sdsm_core.sourceID.buf[i]);
+                }
+            }
+            else{
+                RCLCPP_WARN_STREAM(node_logging_->get_logger(), "Decoded SDSM SourceID does not exist");
+            }
+
+            // EquipmentType
+            if(sdsm_core.equipmentType){
```

Comment:
Can we improve this `if` statement? `sdsm_core.equipmentType` hints that it should be some kind of enumeration, but as typed it looks like a boolean.